python/cuml/common/array.py

```diff
@@ -0,0 +1,269 @@
+
+    from cuml.common.array import Array as cumlArray
+
+    Parameters
+    ----------
+
+    data : rmm.DeviceBuffer, array_like, int
+        An array-like object or integer representing a
+        device or host pointer to pre-allocated memory.
+    owner : object, optional
+        Python object to which the lifetime of the memory
+        allocation is tied. If provided, a reference to this
+        object is kept in this Buffer.
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy or cupy data type.
+    shape : int or tuple of ints, optional
+        Shape of created array.
+    order: string, optional
+        Whether to create a F-major or C-major array.
+
+    Attributes
+    ----------
+
+    ptr : int
+        Pointer to the data
+    size : int
+        Size of the array data in bytes
+    _owner : Python Object
+        Object that owns the data of the array
+    shape : tuple of ints
+        Shape of the array
+    order : {'F', 'C'}
+        'F' or 'C' to indicate Fortran-major or C-major order of the array
+    strides : tuple of ints
+        Strides of the data
+    __cuda_array_interface__ : dictionary
+        __cuda_array_interface__ to interop with other libraries.
+
+    Object Methods
+    --------------
+
+    to_output : Convert the array to the appropriate output format.
+
+    Class Methods
+    -------------
+
+    Array.empty : Create an empty array, allocating a DeviceBuffer.
+    Array.zeros : Create an Array with allocated DeviceBuffer initialized with
+        zeros.
+
+    Notes
+    -----
+
+    cuml Array is not meant as an end-user array library. It is meant for
+    cuML/RAPIDS developer consumption. Therefore it contains the minimum
+    functionality. Its functionality is hidden by base.pyx to provide
+    automatic output format conversion so that the users see the important
+    attributes in whatever format they prefer.
+
+    Todo: support cuda streams in the constructor, particularly after
+
+
+
+    """
+
+    def __init__(self, data=None, owner=None, dtype=None, shape=None,
+                 order=None):
+        if data is None:
+            raise TypeError("To create an empty Array, use the class method" +
+                            " Array.empty()")
+        if isinstance(data, int):
+            if dtype is None or shape is None or order is None:
+                raise TypeError("Need to specify dtype, shape and order when" +
+                                " creating an Array from a pointer.")
+
+        ary_interface = False
+
+        if isinstance(data, DeviceBuffer) or isinstance(data, int):
+            size, shape = _get_size_from_shape(shape, dtype)
+            super(Array, self).__init__(data=data, owner=owner, size=size)
+            self.shape = shape
+            self.dtype = np.dtype(dtype)
+            self.order = order
+            self.strides = _order_to_strides(order, shape, dtype)
```

Comment:
Are there cases where we need more flexibility in `strides` than `order` provides?

Reply:
not currently, but definitely something to keep in mind for the future

Reply:
Just wondering if we can simplify this a bit as `strides` and `order` are somewhat dependent on each other 🙂